Add written proposal for enhancement #103

stringlytyped opened the pull request:
Add written proposal for enhancement #103 which has been adapted from the earlier [_Roadmap to Push Model Support in Keylime_](https://gist.github.com/stringlytyped/367dd6cd29141f4f97b142035203b12a) document

stringlytyped commented:
@THS-on Thanks for giving the PR your approval. I agree that the APIs need a bit of a refactor/clean-up and I think it is natural to tackle that during implementation of the push model. Is anything else needed to get this merged?

stringlytyped commented:
Makes sense. Thanks for clarifying the process

aplanas commented on 103_agent-driven-attestation.md:
```diff
@@ -0,0 +1,421 @@
+
+As part of the current enrolment process, the user specifies a payload which is delivered to the agent and placed in a directory to be consumed by other software. The reason for this is to support the provisioning of identities to workloads running on the node (e.g., TLS certificates or long-lived shared secrets). The payload may optionally contain a script file, which is executed by the agent.
+
+Considering the current landscape of the identity and access management space, a more modern approach to solving this problem would likely be to have Keylime report verification results to a [SPIRE](https://spiffe.io/) attestor plugin which could then handle provisioning of workload identities. This offloads issues related to revocation and suitability for cloud-native workloads.
```
I like this idea. Maybe we can extend it also the the current model?

stringlytyped replied:
To be clear, adding SPIRE integration isn't part of this proposal. There is a separate enhancement (#98) on that, but in light of push support, it may need to be reworked a bit.

aplanas commented on 103_agent-driven-attestation.md:
```diff
@@ -0,0 +1,421 @@
+
+This enhancement proposal requires a relatively significant number of new API endpoints, especially for the verifier. These include those needed to support the new attestation protocol and challenge–response authentication mechanism. Changes to existing endpoints may also be appropriate to expose any new information stored in the database, such as more granular information about the trust status  ...
+
+It suggested that, just as the Rust agent uses [Actix](https://actix.rs/) to serve its REST API, a similar web micro-framework is added as a dependency to the Python codebase. This will provide a natural way to define routes and request handlers and help newcomers understand how the code is organised.
```
what would be the reach of this? like dropping tornado in favor of the new framework?

stringlytyped replied:
You know what, I was basing my understanding of how KL implements its REST APIs on the registrar's code (which uses http.server from the standard library), assuming it was the same globally. However, I see now that the verifier APIs have been implemented using Tornado. Since KL already has a web framework as a dependency, I'll use that naturally and just refactor the existing code a bit

THS-on replied:
Yeah I think we should keep tornado, but should migrate the registrar also to it.

aplanas commented on 103_agent-driven-attestation.md:
```diff
@@ -0,0 +1,421 @@
+
+It suggested that, just as the Rust agent uses [Actix](https://actix.rs/) to serve its REST API, a similar web micro-framework is added as a dependency to the Python codebase. This will provide a natural way to define routes and request handlers and help newcomers understand how the code is organised.
+
+[Flask](https://github.com/pallets/flask) 2.x would be a good candidate for this, as it is packaged for [Fedora since version 37 and EPEL 9](https://packages.fedoraproject.org/pkgs/python-flask/python3-flask/), [Debian Stable](https://packages.debian.org/bookworm/python3-flask), [Ubuntu since version 22.04](https://packages.ubuntu.com/jammy/python3-flask) and [SLES since version 15 SP1](https://s ...
```
should not be something more async oriented, like quart (flask rewrite) or trio?

stringlytyped replied:
I'd have to check the OS support of those, but otherwise I am happy to use any framework. My understanding though is that Tornado is async as well, so we are fine to stick with that, I take it? (see my reply to your previous question)

stringlytyped replied:
The biggest downside of sticking with Tornado is that its router implementation is very bare bones. Quart allows you to specify routers using string templates, which is rather nice and makes the code super easy to read. But not sure if the community would be amenable to switching frameworks

THS-on replied:
I'm fine with switching frameworks if they is well supported and makes our lives easier.

stringlytyped replied:
So, I had a look and quart is not currently packaged by Fedora/EPEL unfortunately (it is in Debian and Ubuntu). That's a shame, because quart's expressiveness would have been nice. Trio seems even more bare bones (its not a web framework), so it probably makes the most sense to stick with Tornado (unless there is some other Python web framework with good async support that I don't know about)

THS-on replied:
There is FastAPI, but it has similar packaging issues to quart.  Yeah sticking with Tornado and cleaning up the common code (parameter extraction, message parsing, responses etc.) is probably the best option for now. Also maybe putting the DB interactions in one single place might make sense to make it more robust.

stringlytyped replied:
Those are all good ideas, thanks Thore. We'll take that approach

aplanas reviewed (commented):
that is a big one, and a very very good one

honestly I still do not have the full picture, but what for the parts that I understand I like very much this path

stringlytyped commented on 103_agent-driven-attestation.md:
```diff
@@ -0,0 +1,421 @@
+
+Once again, we propose two different mechanism for making these trust decisions, which users can use independently or together at their discretion:
+
+- a standard trust store mechanism for use when certificates are available to establish trust; and/or
```
Ah, yeah, the way I've written that is not super clear. KL does currently do some certificate checking, but it is rather limited in scope and can only be used in the specific situations contemplated by KL to date (e.g., for checking EKcerts). The proposal is to genericise this functionality so it can be used for any certificate verification task, including DevIDs and any future certificate-based identity that KL may wish to support. More detail is given in [Design Details: Certificate Trust Store Mechanism](https://github.com/keylime/enhancements/blob/b1791b82100de010ce3b07f5e7c26d2bb44ae0fe/103_agent-driven-attestation.md#certificate-trust-store-mechanism)

I'll clarify this line in the proposal once I've answered your other queries.

stringlytyped commented on 103_agent-driven-attestation.md:
```diff
@@ -0,0 +1,421 @@
+
+3. The agent will gather the information required by the verifier (UEFI log, IMA entries and quote) and report these in a new HTTP request along with other information relevant to the quote (such as algorithms used).
+
+4. The verifier will reply with the number of seconds the agent should wait before performing the next attestation and an indication of whether the request from agent appeared well formed according to basic validation checks. Actual processing and verification of the measurements against policy can happen asynchronously after the response is returned. 
```
If it is too soon, @mheese has suggested that the verifier could reply with a 429 status which I think makes a lot of sense—I just forgot to add it. I'll amend the proposal now to include that detail.

For the "too late" case: I suppose it would make sense to specify a period of validity for the nonce, after which it would expire. If an agent submits an attestation using a nonce that is no longer valid, the verifier would then reply with 400 status. What are your thoughts on this?

aplanas replied:
The "too many requests" and the TTL for the nonce makes a lot of sense :+1:. We can make the TTL > TimeForNextAttestation, so we can play with network delays. This clarify the protocol.

What would be the consideration from the attestation PoV? If there is a 429 I guess that this will not make the note failed in the attestation state (it should be considered trusted), but maybe the 400 should require a change of status if it is repeated?

stringlytyped replied:
To restate all the main error conditions we are considering here:

1. The verifier receives a request before it is expected, in which case it replies with a 429 (too many requests) and the number of seconds to wait in the `Retry-After` header.
2. The verifier receives an attestation for a nonce that is invalid or has expired, in which case it replies with a 400 (bad request).
3. The verifier receives an attestation which violates the policy for the node. In this case, the verifier will reject the next attestation received with a 503 (service unavailable). Of course, once this condition occurs, the verifier will begin reporting the status of the node as having failed attestation when queried.

In the first case, the agent should wait the number of seconds specified by the `Retry-After` header and try again.

In both the second and third case, the agent will retry according to an exponential backoff. This makes sense in the second case as that situation is unexpected behaviour and should only happen in practice, e.g., if the verifier is overloaded and taking a long time to respond to requests.

Exponential backoff is also appropriate in the third case as the user could change the verification policy at any time. As a result, the node cannot simply stop sending attestations when a verification failure occurs. Instead, it will continue to retry periodically in the hopes that an attestation will eventually be accepted.

stringlytyped replied:
To reiterate, a 400 is returned when the verifier receives an attestation for a nonce that is invalid or has expired.

This may occur if the agent or node is behaving unexpectedly. However, it can also occur if the network or the verifier is behaving unexpectedly. E.g., if the verifier is overloading and taking a long time to reply to requests, this could delay the delivery of nonces and cause them to be used outside their period of validity.

So, we cannot use repeated 400s as a reliable indicator of a problem with the node. The verifier should only change the status of the node if an attestation fails verification or if no valid attestation has been received for the configured period of time.

mheese replied:
maybe to add on to that: a successful response should always include a timestamp (or a time duration) on when to query again, so that the agent knows when to perform another request. If it is too early, it receives a 429. (At least that was my idea when we discussed it with @stringlytyped )

stringlytyped replied:
@mheese I agree that for 429s the verifier should include the seconds to wait before trying again (in the `Retry-After` header).

But for 400s, there is no way the verifier can know how long the agent should wait before trying again. Really, if everything is working fine, then the agent should be able to try again straight away by requesting a new nonce.

The 400 condition should not occur under normal operation. So, it makes sense for the agent to wait a while and then try again, increasing the wait time at each attempt until reaching a maximum wait time. This gives the verifier a chance to recover, if it happens that the condition has been caused by an overloaded verifier.

stringlytyped commented:
I've made some changes to the proposal based on the above discussion, specifically:

- @aplanas you noted that you were having a bit of trouble grasping the full picture. I've gone ahead and lifted out the [Agent Lifecycle](https://github.com/keylime/enhancements/blob/0a7c78b151b9f49219ee569fe84516407ccaa779/103_agent-driven-attestation.md#agent-lifecycle) section from my earlier draft and added it to the proposal. If you read through those steps, does that clarify things for you?

- Thinking about it more, I've realised that we do in fact need the nonce to expire. So, I've gone ahead and added more detailed explanations on error handling in [Proposed Attestation Protocol](https://github.com/keylime/enhancements/blob/0a7c78b151b9f49219ee569fe84516407ccaa779/103_agent-driven-attestation.md#proposed-attestation-protocol) (which have also been carried through to the Agent Lifecycle section, mentioned above).

- I've dropped the dependency change.

- Clarified those bits that were confusing.

I noticed the bit about HTTP proxy support was missing from the earlier draft by mistake, so I've added that back in also (see [HTTP Proxy Support](https://github.com/keylime/enhancements/blob/0a7c78b151b9f49219ee569fe84516407ccaa779/103_agent-driven-attestation.md#http-proxy-support)).

maugustosilva commented:
I believe we discussed it at a length and depth that allows the merge and the starting of the actual implementation work. Of course minor course correction might be proposed/request @stringlytyped , but lets not hold it any longer. I again, thank you for the time, effort and thought put on it!
